applications/services/gui/modules/number_input.c

```diff
@@ -0,0 +1,493 @@
+
+/** NumberInput type */
+struct NumberInput {
+    View* view;
+};
+
+typedef struct {
+    const char text;
+    const uint8_t x;
+    const uint8_t y;
+} NumberInputKey;
+
+typedef struct {
+    const char* header;
+    FuriString* text_buffer;
+    bool clear_default_text;
+    int32_t max_value;
+    int32_t min_value;
+    bool useSign;
+    char sign;
+    NumberInputCallback callback;
+    void* callback_context;
+    int8_t selected_row;
+    uint8_t selected_column;
+} NumberInputModel;
+
+static const uint8_t keyboard_origin_x = 7;
+static const uint8_t keyboard_origin_y = 31;
+static const uint8_t keyboard_row_count = 2;
+static const uint8_t enter_symbol = '\r';
+static const uint8_t backspace_symbol = '\b';
+static const uint8_t sign_symbol = '-';
+
+static const NumberInputKey keyboard_keys_row_1[] = {
+    {'0', 0, 12},
+    {'1', 11, 12},
+    {'2', 22, 12},
+    {'3', 33, 12},
+    {'4', 44, 12},
+    {backspace_symbol, 103, 4},
+};
+
+static const NumberInputKey keyboard_keys_row_2[] = {
+    {'5', 0, 26},
+    {'6', 11, 26},
+    {'7', 22, 26},
+    {'8', 33, 26},
+    {'9', 44, 26},
+    {sign_symbol, 55, 17},
+    {enter_symbol, 95, 17},
+};
+
+/** Get row size
+ *
+ * @param      row_index  Index of row
+ *
+ * @return     uint8_t Row size
+ */
+static uint8_t number_input_get_row_size(uint8_t row_index) {
+    uint8_t row_size = 0;
+
+    switch(row_index + 1) {
+    case 1:
+        row_size = COUNT_OF(keyboard_keys_row_1);
+        break;
+    case 2:
+        row_size = COUNT_OF(keyboard_keys_row_2);
+        break;
+    default:
+        furi_crash();
+    }
+
+    return row_size;
+}
+
+/** Get row pointer
+ *
+ * @param      row_index  Index of row
+ *
+ * @return     const NumberInputKey* Row pointer
+ */
+static const NumberInputKey* number_input_get_row(uint8_t row_index) {
+    const NumberInputKey* row = NULL;
+
+    switch(row_index + 1) {
+    case 1:
+        row = keyboard_keys_row_1;
+        break;
+    case 2:
+        row = keyboard_keys_row_2;
+        break;
+    default:
+        furi_crash();
+    }
+
+    return row;
+}
+
+/** Draw input box (common view)
+ *
+ * @param      canvas  The canvas
+ * @param      model   The model
+ */
+static void number_input_draw_input(Canvas* canvas, NumberInputModel* model) {
+    const uint8_t text_x = 8;
+    const uint8_t text_y = 25;
+
+    elements_slightly_rounded_frame(canvas, 6, 14, 116, 15);
+
+    FuriString* text = model->text_buffer;
+    canvas_draw_str(canvas, text_x, text_y, furi_string_get_cstr(text));
+}
+
+static void number_input_backspace_cb(NumberInputModel* model) {
+    uint8_t text_length = model->clear_default_text ? 1 :
+                                                      furi_string_utf8_length(model->text_buffer);
+    if(text_length > 0) {
+        furi_string_set_strn(
+            model->text_buffer, furi_string_get_cstr(model->text_buffer), text_length - 1);
+    }
+}
+
+static char* int32_to_string(int32_t value) {
```

Comment:
FuriString got everything you need, why fallback to C-Strings?

Reply:
Ok, since I always use this to manipulate a FuriString, I'll attempt to use furi_string_printf directly in the function in place of snprintf.